### PR TITLE
Refactor JobsDB data indexing

### PR DESCRIPTION
**Pull request by luk707**
# Motivation

Currently we're indexing items in the database using numerical indexes, I think it makes sense instead to index them by the column name in the database.

# Example

Before:
```python
with closing(sqlite3.connect(self.path)) as connection:
    with closing(connection.cursor()) as cursor:
        connection.execute("PRAGMA foreign_keys = ON")
        cursor.execute(
            """
            SELECT
                id,
                name,
                datetime(created_date,'utc') as created_date
            FROM 
                jobs
            ;
            """,
        )
        for row in cursor.fetchall():
            id = row[0]
            name = row[1]
            created_date = row[2]
```

After:
```python
with closing(sqlite3.connect(self.path)) as connection:
    connection.row_factory = dict_factory
    with closing(connection.cursor()) as cursor:
        connection.execute("PRAGMA foreign_keys = ON")
        cursor.execute(
            """
            SELECT
                id,
                name,
                datetime(created_date,'utc') as created_date
            FROM 
                jobs
            ;
            """,
        )
        for row in cursor.fetchall():
            id = row["id"]
            name = row["name"]
            created_date = row["created_date"]
```

# Summary

This commit introduces a new method called `dict_factory` that implements the [row factory](https://docs.python.org/3/library/sqlite3.html#how-to-create-and-use-row-factories) protocol on the sqlite3 connection class.

I've also implemented `dict_factory` in the JobsDB class and updated the `read_job` method to accept a dictionary representing the database row.

### CODE REVIEW

File: backend/util/dict_factory.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+from sqlite3 import Cursor, Row
+from typing import Dict, Any
+
+
+def dict_factory(cursor: Cursor, row: Row) -> Dict[str, Any]:
+    """
+    This function is used to convert the sqlite3.Row object to a dictionary.
```
Comment (luk707) [inline backend/util/dict_factory.py:7]: Point to the row factory documentation on sqlite.

https://docs.python.org/3/library/sqlite3.html#how-to-create-and-use-row-factories

File: backend/db.py [hunk 2]
```diff
@@ -127,8 +127,8 @@ def update_job(
         job: Job,
     ) -> Job:
         with closing(sqlite3.connect(self.path)) as connection:
+            connection.execute("PRAGMA foreign_keys = ON")
```
Comment (HelinaBerhane) [inline backend/db.py:131]: why was this moved out? also why no `connection.row_factory = dict_factory` here?

File: tests/util/dict_factory_test.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+def mock_database_connection():
+    cursor = MagicMock(spec=sqlite3.Cursor)
+    row = MagicMock(spec=sqlite3.Row)
+
+    cursor.description = [("id", None), ("name", None), ("created_date", None)]
+
+    return cursor, row
+
+
+def test_dict_factory(mock_database_connection):
```
Comment (HelinaBerhane) [inline tests/util/dict_factory_test.py:17]: might be worth having a test with an actual db spun up. could be in memory

Comment (luk707) [reply tests/util/dict_factory_test.py:17]: I think for unit tests we should keep them lean and just ensure that our code calls sqlite as expected.

We can implement E2E tests later.

File: backend/db.py (hunk 2 above)
Comment (luk707) [reply backend/db.py:131]: Theres no need to wait until we have a cursor before we execute the pragma, I'm going to refactor this at a later date as we should setup pooling when dealing with other database types such as postgres.

Comment (HelinaBerhane) [reply backend/db.py:131]: nit: in that case I think move the rest of the `connection.execute("PRAGMA foreign_keys = ON")` lines before the cursor for consistency

**Review by HelinaBerhane — APPROVED** [review]
